File: 1042 Flower Planting With No Adjacent.py
class Solution:
    def gardenNoAdj(self, N: int, paths: List[List[int]]) -> List[int]:
        res = [0] * N
        adj = [[] for i in range(N)]
        for x, y in paths:
            adj[x - 1].append(y - 1)
            adj[y - 1].append(x - 1)
        for i in range(N):
            res[i] = ({1, 2, 3, 4} - {res[j] for j in adj[i]}).pop()
        return res


'''
# faster solution
class Solution:
    def gardenNoAdj(self, N: int, paths: List[List[int]]) -> List[int]:
        ret = [1 for _ in range(N)]
        dc = {}
        for p in paths:
            d1 = min(p)
            d2 = max(p)
            if d2 in dc:
                dc[d2].append(d1)
            else:
                dc[d2] = [d1, ]
            
        for k in sorted(dc.keys()):
            a = []
            for vv in dc[k]:
                a.append(ret[vv-1])
            for x in [1, 2, 3, 4]:
                if x not in a:
                    ret[k-1] = x
                    break
        
        return ret
'''

File: 1042 Flower Planting With No Adjacent.java
class Solution {
    public int[] gardenNoAdj(int N, int[][] paths) {
        Map<Integer, List<Integer>> adj = new HashMap<>();
        for (int i = 0; i < N; i++) adj.put(i, new LinkedList<>());
        for (int[] p : paths) {
            adj.get(p[0] - 1).add(p[1] - 1);
            adj.get(p[1] - 1).add(p[0] - 1);
        }
        int[] res = new int[N];
        for (int i = 0; i < N; i++) {
            int[] colors = new int[5];
            for (int j : adj.get(i))
                colors[res[j]] = 1;
            for (int c = 4; c > 0; --c)
                if (colors[c] == 0)
                    res[i] = c;
        }
        return res;
    }
}